nodeJS/express/controllers.md

```diff
@@ -0,0 +1,460 @@
+1. With an error argument `next(new Error(...))` - Will pass control directly to the error middleware function.
+1. With the string `next('route')` - Will pass control to the next route handler with the same matching path (if there is one). This only works for `app.METHOD` or `router.METHOD`. Potentially, it can also be the same as just calling `next` with no argument.
+1. With the string `next('router')` - Will skip all middleware functions attached to the specific router instance and pass control back out of the router instance. Basically, we exit the router and go back to the parent router, e.g. `app` (yes, the Express app is also just a router under the hood).
+
+Out of the four, you will likely only use the first two, unless you have a very specific need that requires the other two.
+
+### Consolidating what we've learned
+
+A basic example of the file structure would be:
+
+```text
+express-app/
+├─ controllers/
+│  ├─ userController.js
+├─ routes/
+│  ├─ userRoutes.js
+├─ app.js
+```
+
+Creating a user controller file that contains well-defined responsibilities to handle "User".
+
+```javascript
+// user controller file - controllers/userController.js
+
+const asyncHandler = require("express-async-handler");
+
+const getUserById = asyncHandler(async (req, res) => {
+  const userId = req.params.id;
+
+  const user = await someDBQueryToGetUser(userId);
+
+  if (!user) {
+    throw new CustomNotFoundError("User not found");
+  }
+
+  res.send(`User found: ${user.name}`);
+});
+
+const getUsers = asyncHandler(async (req, res) => {
+  // code
+});
+
+const createUser = asyncHandler(async (req, res) => {
+  // code
+});
+
+module.exports = { getUserById, getUsers, createUser };
+```
+
+Route file for the "User"
+
+```javascript
+// route file - routes/userRoutes.js
+
+const express = require('express');
+const userController = require('<path-to-user-controller>');
+
+const router = express.Router();
+
+// You can for example add a top level middleware function that handles say authentication and only let the request come in if they're authenticated
+// This prevents from executing the middleware functions below if the request is not authenticated
+// We will learn more about authentication in later lessons
+// usually calls either next() or next(error)
+
+// router.use(authMiddleware);
+
+
+// router-level midlewares
+
+// router.route allows you to chain route handlers on a specified path
+// will only execute on `/users/` not `/users/a/b/c` for example unless `router.use` is used
+// connect your controllers with the route handlers
+router.route('/')
+  .get(userController.getUsers)
+  .post(userController.createUser);
+  // You will likely place your validation/authentication middleware functions here or perhaps in the controller file, e.g.
+  // .post(validationMiddleware, userController.createUser)
+
+router.route('/:id')
+  .get(userController.getUserById);
+
+module.exports = router;
```

Comment:
can we follow the `router.get("route")` and `router.post("route")` syntax, since that's what we use in other lessons.